### ISSUE TITLE
IndexOutOfRangeException thrown within TextWriterTraceAdapter in async batch event hub triggered function

### ISSUE THREAD

**Issue by Zoidborg7** [Azure/azure-webjobs-sdk#982]
There appears to be a concurrency issue within the TextWriterTraceAdapter that is only revealed under heavy traffic or network delays or some other random cases when processing batches asynchronously. At odd and random times there can be several exceptions per hour, and the exception results in the failure of the triggered webjob function.

We have a few webjobs that use a similar model consuming event hub messages in batches. The log reveals this error occurring in any given webjob at random. At times, the webjobs run without error for many hours, but at other times, the webjobs can experience several exceptions per hour. There are several calls to the TextWriter.Write within the functions and called sub-methods (Write() is called via an extension method I created that prefixes diagnostic info to each output line). The exception can occur either within the initial method called by the webjobs sdk or in any of the methods called within the webjob during the process; again it seems random.

This is the declaration for the event hub triggered webjob function:

```C#
        public static async Task ProcessReceiverEventHubMessagesAsync (
            [EventHubTrigger("ReceiverEventHub")] EventData[] receiverEventDatas,
            CancellationToken cancellationToken,
            TextWriter log);
```

This is the excerpt of log output showing a typical instance of this exception:

```CSV
date,level,applicationName,instanceId,eventTickCount,eventId,pid,tid,message,activityId
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,Exception while executing function: ReceiverProcessor.ProcessReceiverEventHubMessagesAsync,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,Microsoft.Azure.WebJobs.Host.FunctionInvocationException: Exception while executing function: ReceiverProcessor.ProcessReceiverEventHubMessagesAsync ---> System.AggregateException: One or more exceptions thrown during processing. See InnerExceptions for details. ---> System.IndexOutOfRangeException: Index was outside the bounds of the array.,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Text.StringBuilder.get_Chars(Int32 index),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at Microsoft.Azure.WebJobs.Host.Bindings.TextWriterTraceAdapter.Write(Char value),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,"   at System.IO.TextWriter.Write(Char[] buffer, Int32 index, Int32 count)",
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.IO.TextWriter.WriteLine(String value),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,"   at System.IO.TextWriter.WriteLine(String format, Object[] arg)",
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,"   at SharedUtility.IO.TextWriterExtensions.WriteLog(TextWriter textWriter, String formatString, Object[] args)",
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at ReceiverMessageJob.ReceiverProcessor.<BeginReceiverProcessingAsync>d__11.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at ReceiverMessageJob.ReceiverProcessor.<ProcessAsync>d__10.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at ReceiverMessageJob.ReceiverProcessor.<ProcessReceiverEventHubMessagesAsync>d__0.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   --- End of inner exception stack trace ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at ReceiverMessageJob.ReceiverProcessor.<ProcessReceiverEventHubMessagesAsync>d__0.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`1.<InvokeAsync>d__0.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<InvokeAsync>d__53.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithWatchersAsync>d__3b.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__36.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__1a.MoveNext(),
2017-01-18T21:10:12,Information,my-application-name,6d3b73,636203706125097178,0,1692,12,   --- End of inner exception stack trace ---,
```

I looked into the SDK code indicated at the top of the exception stack trace where it seems the exception is originating from, and I found this in https://github.com/Azure/azure-webjobs-sdk/blob/master/src/Microsoft.Azure.WebJobs.Host/Bindings/TraceWriter/TextWriterTraceAdapter.cs:

```C#
31:        public override void Write(char value)
32:        {
33:            // buffer all output
34:            _text.Append(value);
35:
36:            int len = _text.Length;
37:            if (len > 2 && _text[len - 2] == '\r' && _text[len - 1] == '\n')
38:            {
39:                // when we see a newline, flush the output
40:                // flushing often is very important - we need to ensure that output
41:                // is flushed to the Dashboard as it is written in the body of
42:                // long running executing functions.
43:                _traceWriter.Info(_text.ToString());
44:                _text.Clear();
45:            }
46:        }
```

This is speculation, but my best guess is that somehow the same StringBuilder instance (_text) is being cleared (perhaps in line 44) while its previous Length value (len assigned at line 36) is being used to index the same StringBuilder at line 37. The timing would have to be perfect for the _text to be cleared between lines 36 and 37, but it is definitely possible without a lock or semaphore in use. Again, this is my best guess; I wasn't able to observe the problem happening while debugging.

Our event hubs get messages every second from the source devices, and the indicated method where the exception originates (ReceiverMessageJob.ReceiverProcessor.BeginReceiverProcessingAsync()) is called within threads for each of the event hub messages contained within the batch. I suppose I could implement semaphore access to the log TextWriter within my application, but that isn't something you normally expect to have to do for a log TextWriter is it?

For now I have instituted a work-around. I have bypassed the use of the Dashboard-based TextWriter provided to the triggered function method by assigning Console.Out to the TextWriter parameter provided to the triggered function method (as seen below). This has eliminated the exception entirely.

```C#
        public static async Task ProcessReceiverEventHubMessagesAsync (
            [EventHubTrigger("ReceiverEventHub")] EventData[] receiverEventDatas,
            CancellationToken cancellationToken,
            TextWriter log)
        {
            // Use Console instead of WebJobs dashboard for log output.
            log = Console.Out;
            ...
```

I can get to the output using the diagnostic log for the Azure App Service's diagnostic logs sent to Blob storage (which is where the log excerpt above is obtained from in the first place; now that log contains all output). However, the webjob dashboard function call instances will contain no output as a result of this work-around.

We are using the 2.0.0-beta2 nuget package release of the webjobs SDK and the webjobs servicebus SDK, but all other referenced nuget packages are latest release as of 1/15/2017.

**Comment by brettsam** [comment]
At first glance, it looks like the same problem with StringBuilder that we addressed here: https://github.com/Azure/azure-webjobs-sdk/pull/849

**Comment by Zoidborg7** [comment]
That is possible. I went to the referenced issue #675 and searched my log for a similar exception and found it (ArgumentOutOfRangeException instead of IndexOutOfRangeException). The stack trace looks very similar and comes out of the same method (Microsoft.Azure.WebJobs.Host.Bindings.TextWriterTraceAdapter.Write(Char value)):

```CSV
date,level,applicationName,instanceId,eventTickCount,eventId,pid,tid,message,activityId
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,Executed: 'ReceiverProcessor.ProcessReceiverEventHubMessagesAsync' (Failed),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,Microsoft.Azure.WebJobs.Host.FunctionInvocationException: Exception while executing function: ReceiverProcessor.ProcessReceiverEventHubMessagesAsync ---> System.AggregateException: One or more exceptions thrown during processing. See InnerExceptions for details. ---> System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,Parameter name: chunkLength,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Text.StringBuilder.ToString(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Bindings.TextWriterTraceAdapter.Write(Char value),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,"   at System.IO.TextWriter.Write(Char[] buffer, Int32 index, Int32 count)",
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.IO.TextWriter.WriteLine(String value),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,"   at System.IO.TextWriter.WriteLine(String format, Object[] arg)",
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,"   at SharedUtility.IO.TextWriterExtensions.WriteLog(TextWriter textWriter, String formatString, Object[] args)",
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at ReceiverMessageJob.ReceiverProcessor.<BeginReceiverProcessingAsync>d__11.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at ReceiverMessageJob.ReceiverProcessor.<ProcessAsync>d__10.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at ReceiverMessageJob.ReceiverProcessor.<ProcessReceiverEventHubMessagesAsync>d__0.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   --- End of inner exception stack trace ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at ReceiverMessageJob.ReceiverProcessor.<ProcessReceiverEventHubMessagesAsync>d__0.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`1.<InvokeAsync>d__0.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<InvokeAsync>d__53.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithWatchersAsync>d__3b.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__36.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__1a.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   --- End of inner exception stack trace ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<ExecuteWithLoggingAsync>d__1a.MoveNext(),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,--- End of stack trace from previous location where exception was thrown ---,
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task),
2017-01-18T21:13:50,Information,my-application-name,6d3b73,636203708306106140,0,1692,13,   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.<TryExecuteAsync>d__2.MoveNext(),
```

**Comment by Zoidborg7** [comment]
Just to be clear @brettsam : is there something that I need to do relating to the previously addressed issue you mentioned, or is that other issue mentioned as being the same problem but in a previously unidentified place (the TextWriterTraceAdapter)? Thanks!

**Comment by mathewc** [comment]
Looks like this might be another related issue to the one @brettsam fixed in the fix he referenced above.

**Comment by brettsam** [comment]
@Zoidborg7 -- I just noticed that you're using a `TextWriter` to write logs. What if you used `TraceWriter`? It looks like a pretty common scenario so I was trying to think why we're not seeing this more often (it's definitely a bug). My guess is that most people use `TraceWriter` to do their logging now, which shouldn't have this problem.

You could replace your calls to `log.Write()` with `log.Info()`.

**Comment by brettsam** [comment]
@mathewc -- is `TextWriter` support for logging still around for legacy purposes, or is there a good reason to keep it? Since there's a bug here, I'm wondering if we just remove support and require use of `TraceWriter`.

**Comment by mathewc** [comment]
I saw no real reason to remove the earlier TextWriter logging capability so left it in place when I added TraceWriter.

**Comment by Zoidborg7** [comment]
@brettsam -- I was not aware of the TraceWriter. I'm assuming that the usage is through the function's parameters (same as the TextWriter parameter)? I probably won't have an opportunity to try that out for at least a few weeks due to our current workload, but I will look into it. Thanks!

Should the TextWriter be updated or deprecated or perhaps even be changed to just be a wrapper for the TraceWriter for backward compatibility?

**Comment by brettsam** [comment]
Fixed with #1043

**Comment by Zoidborg7** [comment]
So with the fix #1043 I should be able to use the TextWriter for the log output? It would be preferable to having to switch to TraceWriter since the other log output methods I have used in the meantime are also TextWriter derived classes and therefore required no modification to the many calls to Write() etc.

**Comment by brettsam** [comment]
Yes, TextWriter is supported.
